src/storage/mvcc/reader/forward_scanner.rs

```diff
@@ -0,0 +1,428 @@
+        let mut needs_seek = true;
+
+        for _ in 0..SEEK_BOUND {
+            {
+                let current_key = self.write_cursor.key(&mut self.statistics.write);
+                if !Key::is_user_key_eq(current_key, user_key.encoded().as_slice()) {
+                    // Meet another key.
+                    return Ok(None);
+                }
+                if Key::decode_ts_from(current_key)? <= safe_ts {
+                    // Founded, don't need to seek again.
+                    needs_seek = false;
+                    break;
+                }
+            }
+            self.write_cursor.next(&mut self.statistics.write);
+            if !self.write_cursor.valid() {
+                // Key space ended.
+                return Ok(None);
+            }
+        }
+        // If we have not found `${user_key}_${safe_ts}` in a few `next()`, directly `seek()`.
+        if needs_seek {
+            self.write_cursor.seek(
+                &user_key.clone().append_ts(safe_ts),
+                &mut self.statistics.write,
+            )?;
+            if !self.write_cursor.valid() {
+                // Key space ended.
+                return Ok(None);
+            }
+            let current_key = self.write_cursor.key(&mut self.statistics.write);
+            if !Key::is_user_key_eq(current_key, user_key.encoded().as_slice()) {
+                // Meet another key.
+                return Ok(None);
+            }
+        }
+
+        // Now we must have reached the first key >= `${user_key}_${safe_ts}`. However, we may
+        // meet `Lock` or `Rollback`. In this case, more versions needs to be looked up.
+        loop {
+            let write = Write::parse(self.write_cursor.value(&mut self.statistics.write))?;
+            self.statistics.write.processed += 1;
+
+            match write.write_type {
+                WriteType::Put => return Ok(Some(self.load_data_by_write(write, user_key)?)),
+                WriteType::Delete => return Ok(None),
+                WriteType::Lock | WriteType::Rollback => {
+                    // Continue iterate next `write`.
+                }
+            }
+
+            self.write_cursor.next(&mut self.statistics.write);
+
+            if !self.write_cursor.valid() {
+                // Key space ended.
+                return Ok(None);
+            }
+            let current_key = self.write_cursor.key(&mut self.statistics.write);
+            if !Key::is_user_key_eq(current_key, user_key.encoded().as_slice()) {
+                // Meet another key.
+                return Ok(None);
+            }
+        }
+    }
+
+    /// Load the value by the given `write`. If value is carried in `write`, it will be returned
+    /// directly. Otherwise there will be a default CF look up.
+    ///
+    /// The implementation is the same as `PointGetter::load_data_by_write`.
+    #[inline]
+    fn load_data_by_write(&mut self, write: Write, user_key: &Key) -> Result<Value> {
```

Comment:
should we extract a trait for these readers?

Reply:
Cannot. It modifies self.default_cursor (may create new one with different configurations, i.e. in forward scanner, create a forward cursor and in backward scanner, create a backward cursor)